pkg/master/publish.go

```diff
@@ -147,11 +151,53 @@ func (m *Master) setEndpoints(serviceName string, ip net.IP, port int) error {
 			},
 		}
 	}
-	if !reflect.DeepEqual(e.Subsets, want) {
-		e.Subsets = want
-		glog.Infof("setting endpoints for master service %q to %v", serviceName, e)
+
+	// First, determine if the endpoint is in the format we expect (one
+	// subset, one port, N IP addresses).
+	formatCorrect, foundIP := m.checkEndpointSubsetFormat(e, ip.String(), port)
+	if !formatCorrect {
+		// Something is egregiously wrong, just re-make the endpoints record.
+		e.Subsets = []api.EndpointSubset{{
+			Addresses: []api.EndpointAddress{{IP: ip.String()}},
+			Ports:     []api.EndpointPort{{Port: port, Protocol: api.ProtocolTCP}},
+		}}
+		glog.Infof("Resetting endpoints for master service %q to %v", serviceName, e)
+		return m.endpointRegistry.UpdateEndpoints(ctx, e)
+	} else if !foundIP {
+		// So we *always* add our own IP address to the end, and if
+		// there are too many IP addresses, we remove from the
+		// beginning. Given the requirements stated at the top of this
+		// function, this should cause the list of IP addresses to
```

Comment:
The sorting is done in the RepackSubsets function.

On Tue, Apr 28, 2015 at 5:33 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/master/publish.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7273#discussion_r29302403>
> :
>
> > +	// First, determine if the endpoint is in the format we expect (one
> > +	// subset, one port, N IP addresses).
> > +	formatCorrect, foundIP := m.checkEndpointSubsetFormat(e, ip.String(), port)
> > +	if !formatCorrect {
> > +		// Something is egregiously wrong, just re-make the endpoints record.
> > +		e.Subsets = []api.EndpointSubset{{
> > +			Addresses: []api.EndpointAddress{{IP: ip.String()}},
> > +			Ports:     []api.EndpointPort{{Port: port, Protocol: api.ProtocolTCP}},
> > +		}}
> > +		glog.Infof("Resetting endpoints for master service %q to %v", serviceName, e)
> > +		return m.endpointRegistry.UpdateEndpoints(ctx, e)
> > +	} else if !foundIP {
> > +		// So we *always* add our own IP address to the end, and if
> > +		// there are too many IP addresses, we remove from the
> > +		// beginning. Given the requirements stated at the top of this
> > +		// function, this should cause the list of IP addresses to
>
> I will make it sort and then delete the address(es) that come after it's
> own address. I think that has faster convergence properties than deleting a
> random one.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7273/files#r29302403>
> .
>